### PR TITLE
core: (trait) Implement LazyTrait to workaround circular dependency

### PR DESCRIPTION
**Pull request by kingiler**
Address #1218 

`LazyTrait` is implemented to postpone evaluation of provided traits.

### CODE REVIEW

**Review by PapyChacal — COMMENTED** [review]
Oh wow, this still looks hacky, but is somehow minimally intrusive I guess?
I'm all for it personally, I'd rather have a functional hack in 🙂

That said, this makes me want to try wrapping it in a utility function though, which could show intent
Say `HasParent(forward(scf.For))`; do you think it would work? Also wondering What others think

Either way, thanks a lot!

**Review by PapyChacal — APPROVED** [review]

**Comment by PapyChacal** [comment]
And I think it could work at the OpTrait level rather than a specific trait?

**Review by math-fehr — COMMENTED** [review]
Nice!
I'm wondering if it would be better to do that directly in `traits`?
So that way, we get that for all traits for free, instead of having it here only for the `HasParent` trait?

File: xdsl/dialects/scf.py [hunk 1]
```diff
@@ -78,10 +78,12 @@ class Yield(IRDLOperation):
     name = "scf.yield"
     arguments: VarOperand = var_operand_def(AnyAttr())
 
-    # TODO circular dependency disallows this set of traits
-    # tracked by gh issues https://github.com/xdslproject/xdsl/issues/1218
-    # traits = frozenset([HasParent((For, If, ParallelOp, While)), IsTerminator()])
-    traits = frozenset([IsTerminator()])
+    traits = frozenset(
+        [
+            HasParent((lambda: For, lambda: If, lambda: ParallelOp, lambda: While)),
```
Comment (superlopuh) [inline xdsl/dialects/scf.py:83]:
```suggestion
            HasParent((lambda: (For, If, ParallelOp, While))),
```

Comment (superlopuh) [reply xdsl/dialects/scf.py:83]: Would the following work? It feels more readable to me

**Review by superlopuh — COMMENTED** [review]
This will work, but I wonder if we would benefit from a more general mechanism to separate the logic. I'm thinking something like a `LazyTrait(lambda: AnyTraitYouWant())`, which would let us test the logic separately, reuse it for other circular dependencies, and avoid the frozen workaround. Could you please give it a go and check whether it would actually simplify the code a bit?

**Comment by superlopuh** [comment]
Ah I see others had the same idea :)

**Comment by kingiler** [comment]
Do you want all traits to be lazy? We can have the implementation allows:
```python3
    traits = frozenset(
        LazyTrait(
            lambda: (
                IsolatedFromAbove(),
                SingleBlockImplicitTerminator(ModuleEndOp),
                SymbolOpInterface(),
                SymbolTable(),
            )
        )
    )
```

File: xdsl/traits.py [hunk 2]
```diff
@@ -41,15 +43,31 @@ class HasParent(OpTrait):
...
+        if self._unresolved:
+            object.__setattr__(
+                self,
+                "parameters",
+                tuple(
+                    parameter if isclass(parameter) else parameter()
+                    for parameter in self.parameters
+                ),
+            )
+            object.__setattr__(self, "_unresolved", True)
```
[the remark this answers is not in the document]
Comment (kingiler) [reply xdsl/traits.py]: No longer have hacky ways in newest commit.

**Pull request closed without merging**